variables.go

```diff
@@ -123,10 +123,60 @@ func (e *Executor) compiledTask(call taskfile.Call, evaluateShVars bool) (*taskf
 			if cmd == nil {
 				continue
 			}
+			if cmd.ForEach != nil {
+				var list []string
+				// Get the list from the explicit foreach list
+				if cmd.ForEach.List != nil && len(cmd.ForEach.List) > 0 {
+					list = cmd.ForEach.List
+				}
+				// Get the list from the task sources
+				if cmd.ForEach.From == "source" {
+					list, err = fingerprint.Globs(new.Dir, new.Sources)
```

Comment:
I noticed that `foreach: source` is returning absolute paths like:

```
/Users/andrey/Developer/andrey/task/args/args.go
```

instead relative like

```
args/args.go
```

We need to discuss which would be more appropriate to the user: absolute or relative paths. Ideally, whichever option we choose, the user should be able to convert one into another. I'm inclined to prefer relative. We can use [`filepath.Rel`](https://pkg.go.dev/path/filepath#Rel) to convert an abs path to relative.

What may become tricky here is that we should take into account _the task directory_ and not the user directory when converting.

Maybe we should add a `TASK_DIR` special var similar to [the other special variables](https://taskfile.dev/api/#special-variables) we already have. Then, the user would be able to do something like `echo {{join .TASK_DIR .ITEM}}` to get the absolute path.

(This `join` function doesn't seem to be included into [slim-sprig](https://go-task.github.io/slim-sprig/paths.html) nor on the list of functions [we expose ourselves to the user](https://github.com/go-task/task/blob/main/internal/templater/funcs.go). We will need to add it. Given a `rel` function is also missing, we could probably add it as well.)

Any thoughts?

Reply:
This is a super interesting point that I didn't consider. I completely agree that having functions to switch between relative (`rel`) and absolute (`abs`) URLs would be really useful (and also a `join`).

However, I'm not at all sure which type of path would be most useful to the user. Out of curiosity, is there any reason that you lean towards relative over absolute?

Reply:
If feels "cleaner", I guess. Less verbose.

There's no right on wrong on this decision, though.

I think I'm fine if we keep as absolute as well, but with the proper function to convert + an example on the docs on how to do it.

Reply:
@pd93 

> You could make a function that accepts the current directory as a parameter.

I think this route would be an appropriate middle ground. What do you think?

I just looked at the Go's documentation and realized that [`filepath.Rel`](https://pkg.go.dev/path/filepath#Rel) actually already asks for the base path, so we asking on our side would be consistent.

Regarding `abs`, it would actually just be a `pathJoin(<working-dir>, <path>)`, so maybe we could just skip `abs` to something we would consider to do in the future.

The important here is to properly document this in the documentation so it's easy enough for users to find and use.

---

Last thought, maybe we should add `path` to all functions for clarity:

- `joinPath`
- `relPath`
- `absPath` (future)

Reply:
I have added the `joinPath` and `relPath` functions. I wasn't sure if these should be in slim-sprig or Task, so I've just added them to Task for now.

I did also look into adding a `TASK_DIR` special variable, but this is a bit awkward at the moment as `t.Dir` is not resolved until after the variable compilation happens and `t.Dir` also accepts variables which need to be resolved. There are ways to solve this, but I think doing lazy variables (in some form) will solve that issue anyway, so maybe this is another future thing.

I've also added some docs to the `for` section describing how you would go about converting the relative paths to absolute ones.

Reply:
> I have added the `joinPath` and `relPath` functions. I wasn't sure if these should be in slim-sprig or Task, so I've just added them to Task for now.

That's fine. We can always move them to `slim-sprig` later if we want to.